bgzf/reader.go

```diff
@@ -4,22 +4,84 @@ import (
 	"bufio"
 	"bytes"
 	"compress/gzip"
+	"io"
 	"log"
 	"math"
 	"os"
 )
 
-// Reader is the decompressor for BGZF files which is at it's core
+// BlockReader is the decompressor for BGZF files which is at its core
 // a gzip reader. The BGZF reader retains a *os.File to use for
 // seek functionality as well as a *bufio.Reader to implement the
 // io.ByteReader interface, required for reading gzip files as
 // independent concatenated blocks, rather than a multistream.
-type Reader struct {
+type BlockReader struct {
 	file         *os.File
 	buf          *bufio.Reader
 	decompressor *gzip.Reader
 }
 
+// Read wraps a BlockReader and provides the Next and Read functions
+// that automatically read a new block when the previous block has
+// been completely read.
+type Reader struct {
+	br           *BlockReader
+	blk          *Block
+	intermediate bytes.Buffer
+	eof          bool
+}
+
+// Next returns the next n bytes from the bgzf block.
+// Next handles reading new block if more bytes are
+// requested than are in the currently store block.
+func (r *Reader) Next(n int) []byte {
```

Comment:
Can this be written in procedural style instead?
Next(r *Reader, n int)
Is there an advantage to having it be a method?